### ISSUE TITLE
Publish does not update package-lock

### ISSUE THREAD

**Issue by Robinfr** [lerna/lerna#1415]
## Expected Behavior
When doing a publish, the version number should not only be bumped in the `package.json` but also in the `package-lock.json` when available.

## Current Behavior
The package-lock version does not get updated. This causes the version to be updated the next time an install is run creating an extra (useless) commit.

## Possible Solution
`package-lock.json` version number should be made equal to the `package.json` version number when publishing. 

## Context
I'm doing a Lerna publish automatically. This pulls the latest version and publishes all of the packages to NPM. However, now I constantly have to create extra commits for the package-lock.

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.2.0 |
| `npm --version`  | 5.8.0 |
| `node --version` | 8.11.1 |

**Comment by evocateur** [comment]
The root `package-lock.json`, or the leaf package-locks (`packages/*/package-lock.json`)?

**Comment by Robinfr** [comment]
The leaf package locks.

Op 8 mei 2018 6:27 p.m. schreef Daniel Stockman <notifications@github.com>:

The root package-lock.json, or the leaf package-locks (packages/*/package-lock.json)?

**Comment by evocateur** [comment]
We're not calling `npm version`, just modifying the `package.json` directly and then calling `npm publish`. I try to avoid leaf package lockfiles, much like lerna itself does (requires lerna 3.x).

**Comment by evocateur** [comment]
> Is there any reason not to?

For lerna 2.x, it's because I don't want to deal with that old codebase anymore.

For lerna 3.x (current `master`), it's a bandwidth issue. I'm only one person, and I have a day-job.

**Comment by Robinfr** [comment]
I'm not asking whether there is a reason for you not to, I'm asking in general. One of the pros of open source is that you don't have to be solely responsible for developing the software. If you agree with the approach I could try taking a look at implementing it when I have some time to help you out.

Op 8 mei 2018 9:29 p.m. schreef Daniel Stockman <notifications@github.com>:

Is there any reason not to?

For lerna 2.x, it's because I don't want to deal with that old codebase anymore.

For lerna 3.x (current master), it's a bandwidth issue. I'm only one person, and I have a day-job.

**Comment by evocateur** [comment]
Sure, a PR would be great.

**Comment by tizmagik** [comment]
I'd be happy to assist with this as well if you could point me in the right direction 😄

**Comment by Robinfr** [comment]
I was reading through the code but I think I also need some directions, I understand how the system works but it's entirely centered around the `package.json` file so adding the code for `package-lock.json` anywhere feels hacky to me..

@evocateur how did you disable the `package-lock.json` in the Lerna packages anyway? I can't find any `.npmrc` file or `.gitignore` in the project that would do this.

**Comment by doug-wade** [comment]
@Robinfr we use [npm's `file:` specifiers](https://github.com/npm/npm/pull/15900) to manage our monorepo, which means that when you run `bootstrap`, `npm i` is only run in the root directory.  Thus, the extraneous `package-lock.json`s are never generated.

**Comment by Robinfr** [comment]
@doug-wade but the additional dependencies in the sub-packages are not installed?

**Comment by garthk** [comment]
@Robinfr when Lerna writes its temporary `package.json` it pulls out everything it thinks it doesn't need — including `version` and `name`, both of which get materialised into `package-lock.json` in the way you need. I think the fix is to write both `version` and `name` to the temporary `package.json`. If you're doing `version`, could you also do `name` for #1454? Thanks.

**Comment by garthk** [comment]
@Robinfr `npm` derives `package-lock.json` from `package.json`, so by changing how Lerna produces `package.json` during `bootstrap` you can get the desired result observed in `package-lock.json` without having to write code to read, alter, and re-write the latter.

**Comment by evocateur** [comment]
publish and bootstrap are completely separate operations, and bootstrap should not elicit unstaged changes. thus it needs to be updated during `lerna publish`. but it is not, at present, because i personally never use nested lockfiles.

**Comment by tivac** [comment]
I'm currently working around this by running

```bash
> npx lerna exec -- npm i
```

after each `lerna publish`. It's not ideal but it at least gets things stabilized again.

@evocateur what's your argument against nested lockfiles? I've got all my `devDependencies` hoisted to the root, but I still want to have consistent installs of the regular `dependencies` for everything in my packages dir.

I'm currently using [`npm ci` for installs](https://github.com/tivac/modular-css/blob/6ce23fa8def8cf46a2c40b772b42d3397fae6c49/.travis.yml#L30) and I definitely don't want to let it just randomly decide to upgrade a nested package.

**Comment by tivac** [comment]
I'm using relative file specifiers in `modular-css`, but I've still got `dependencies` in all my packages because I can't hoist those to the root. Lerna looks to be doing the same thing?

https://github.com/lerna/lerna/blob/f674f354f0260c57d885c181e16c9ce23ac252a8/commands/diff/package.json#L38

So I still don't get how you get around having leaf node `package-lock.json` files. I've got one in the root for all my `devDependencies` just like lerna, but that doesn't help w/ anything in `packages/` does it?

Would definitely love to know how I could improve my setup or workflow for using `lerna@3` in `modular-css` so I appreciate any wisdom you can share!

**Comment by lizardruss** [comment]
> Relative file: specifiers, just like the lerna source. It’s a unified package-lock for the entire tree, no bootstrap required. (I really need to document this pattern, I feel like I’ve repeated myself about eleventy-billion times over the past six months)

This approach won't work when a leaf package is going to be built into a Docker image. The `docker build` process won't follow symlinks, and the `file:///` specifiers won't resolve unless the local dependencies are copied into the build context.

I'm not requesting a fix, only giving this example if anyone runs into the same issue I'm having. I realize it's not within the scope of this project to support docker builds.

**Comment by trippingtarballs** [comment]
My work around ...

## Fix

Add the following to the root `package.json` of your project:

```json
{
  "scripts": {
    "version": "lerna clean --yes && lerna bootstrap --ignore-scripts -- --package-lock-only --no-audit && git add packages/*/package-lock.json"
  }
}
```

## Explanation

When the `lerna publish` command is executed the `version` run-script is invoked by `lerna` [_"AFTER bumping the package version, but BEFORE commit"_](https://github.com/lerna/lerna/blob/5a7392e59e6ae5ee8aea10467ffb4b12f3b1e3d1/commands/version/index.js#L408). So, breaking down the above run-script:

* `lerna clean --yes`
  * This is necessary because if a `node_modules` folder is present during the installation of a package's dependencies then the version that package's `package-lock.json` will not be updated by `npm`. I claim as much based on trial and error.
  * Given my workflow (where I perform a `lerna publish`) I do not expect there to be a `node_modules` present for any of my project's packages. However, it does no harm to be explicit.
* `lerna bootstrap --ignore-scripts -- --package-lock-only --no-audit`
  * The use of `lerna bootstrap` over `lerna exec "npm install"` is because that is [what the `bootstrap` command is for](https://github.com/lerna/lerna/tree/v3.10.6/commands/bootstrap#usage).
  * The use of the `--ignore-scripts` option is because I use `preinstall` & `postinstall` in my project's root `package.json` and, for my particular `lerna publish` workflow, I do not care to have those run-scripts get triggered.
  * The use of the `--ignore-scripts` option before the `--` is to instruct both `bootstrap` and `npm` to not execute any run-scripts.
  * `--package-lock-only` [_"argument will only update the `package-lock.json`, instead of checking `node_modules` and downloading dependencies"_](https://docs.npmjs.com/cli/install).
  * `--no-audit` again, another time saver. I really just want to bump the version in the package's `package-lock.json`.
* `git add packages/*/package-lock.json`
  * Specific to my use-case. I have all my project's packages as subfolders of `packages/`.
  * The use of the `*` in `packages/*/package-lock.json` means the `git add` will only match on the `package-lock.json` file that is in a folder one-level down. Again, something that makes sense for my project.

## Possible PR?

The need for a `git add` at all is due to the following lines (from `lerna` source code, [`commands/version/index.js`](https://github.com/lerna/lerna/blob/v3.10.6/commands/version/index.js#L496-L501)):

```js
    // exec version lifecycle in root (after all updates)
    chain = chain.then(() => this.runRootLifecycle("version"));

    if (this.commitAndTag) {
      chain = chain.then(() => gitAdd(Array.from(changedFiles), this.execOpts));
    }
```

The logic of the `version` command for `lerna` will; 1. compile a list of files that are to be included with the tag commit, 2. execute whatever `version` run-script you might have in your project's root `package.json`, and 3. execute the `git add` part of the tag commit.

In other words, it's not enough that `lerna bootstrap --ignore-scripts -- --package-lock-only --no-audit` would have updated the `package-lock.json` because `lerna` would have already decided on what files to stage for the tag commit. Hence why I include `git add packages/*/package-lock.json` – to ensure the `package-lock.json` is added to the tag commit.

## Conclusions

Apologies for the monologue. This issue has been raised more than once (#925). Hopefully my input is helpful to some.

I did think about a PR to have the `git add` part of the tag commit be moved to after the `version` run-script has been invoked. But I don't know enough about `lerna`'s source code to know if such a change would have [unwanted side-effects](https://github.com/abraham/nutmeg/pull/584).

**Comment by evocateur** [comment]
@trippingtarballs The `git add` in your custom lifecycle script is perfectly appropriate. Just because you're using Lerna doesn't mean Lerna must do _all_ the work.

**Comment by binomialstew** [comment]
@trippingtarballs, in your example, when/how does the commit happen for the lock files? If I understand correctly, this is supposed to happen in the same commit where versions are tagged, but if I have these changed lock files added during the version script, I see the fatal `lerna ERR! EUNCOMMIT Working tree has uncommitted changes`.

**Comment by huboneo** [comment]
This full on breaks NPM functionality for lerna packages. Since dependency changes are not reflected in lock files basic npm commands such as `npm audit` fail to execute

**Comment by matthias-ccri** [comment]
If you're using `hoist`, the `lerna exec "npm i"` workaround is not a good solution, because it generates a bad lockfile. Normally when bootstrapping with hoisting, lerna populates the lockfile with _only_ the non-hoisted packages. `npm i` populates it will _all_.

Then, when you run `lerna bootstrap`, the lockfile takes precedence. Npm installs from the lockfile and not from the package.json, which effectively means packages aren't hoisted properly. I may be wrong about that, but it's what I've seen.

**Comment by remjx** [comment]
This workaround ensures that leaf packages are included as dependencies in the package-locks of other leaf packages: https://github.com/lerna/lerna/issues/1462#issuecomment-638684240
